File: js/src/lib/output-blocks.js
/**
 * Output formatting utilities for nicely rendered command blocks
 *
 * Provides "timeline" format: a width-independent, lossless output format
 * that works in TTY, tmux, SSH, CI, and logs.
 *
 * Core concepts:
 * - `│` prefix → tool metadata (timeline marker)
 * - `$` → executed command (virtual or user command)
 * - No prefix → program output (stdout/stderr)
 * - Result marker (`✓` / `✗`) appears after output
 */

// Timeline marker character (formerly called "spine")
const TIMELINE_MARKER = '│';

// Alias for backward compatibility
const SPINE = TIMELINE_MARKER;

// Result markers
const SUCCESS_MARKER = '✓';
const FAILURE_MARKER = '✗';

/**
 * Create a metadata line with timeline marker prefix
 * @param {string} label - Label (e.g., 'session', 'start', 'exit')
 * @param {string} value - Value for the label
 * @returns {string} Formatted line with timeline marker prefix
 */
function createTimelineLine(label, value) {
  // Pad label to 10 characters for alignment
  const paddedLabel = label.padEnd(10);
  return `${TIMELINE_MARKER} ${paddedLabel}${value}`;
}

/**
 * Alias for backward compatibility
 * @deprecated Use createTimelineLine instead
 */
function createSpineLine(label, value) {
  return createTimelineLine(label, value);
}

/**
 * Create an empty timeline line (just the timeline marker character)
 * @returns {string} Empty timeline line
 */
function createEmptyTimelineLine() {
  return TIMELINE_MARKER;
}

/**
 * Alias for backward compatibility
 * @deprecated Use createEmptyTimelineLine instead
 */
function createEmptySpineLine() {
  return createEmptyTimelineLine();
}

/**
 * Create a command line with $ prefix
 * @param {string} command - The command being executed
 * @returns {string} Formatted command line
 */
function createCommandLine(command) {
  return `$ ${command}`;
}

/**
 * Get the result marker based on exit code
 * @param {number} exitCode - Exit code (0 = success)
 * @returns {string} Result marker (✓ or ✗)
 */
function getResultMarker(exitCode) {
  return exitCode === 0 ? SUCCESS_MARKER : FAILURE_MARKER;
}

/**
 * Create a virtual command block for setup steps (like docker pull)
 * Virtual commands are displayed separately in the timeline to show
 * intermediate steps that the tool performs automatically
 * @param {string} command - The virtual command being executed
 * @returns {string} Formatted command line
 */
function createVirtualCommandBlock(command) {
  return createCommandLine(command);
}

/**
 * Create a result marker line for a virtual command
 * @param {boolean} success - Whether the virtual command succeeded
 * @returns {string} Result marker (✓ or ✗)
 */
function createVirtualCommandResult(success) {
  return success ? SUCCESS_MARKER : FAILURE_MARKER;
}

/**
 * Create a separator line between virtual commands and user commands
 * @returns {string} Empty timeline line
 */
function createTimelineSeparator() {
  return createEmptyTimelineLine();
}

/**
 * Parse isolation metadata from extraLines
 * Extracts key-value pairs from lines like "[Isolation] Environment: docker, Mode: attached"
 * @param {string[]} extraLines - Extra lines containing isolation info
 * @returns {object} Parsed isolation metadata
 */
function parseIsolationMetadata(extraLines) {
  const metadata = {
    isolation: null,
    mode: null,
    image: null,
    container: null,
    screen: null,
    session: null,
    endpoint: null,
    user: null,
  };

  for (const line of extraLines) {
    // Parse [Isolation] Environment: docker, Mode: attached
    const envModeMatch = line.match(
      /\[Isolation\] Environment: (\w+), Mode: (\w+)/
    );
    if (envModeMatch) {
      metadata.isolation = envModeMatch[1];
      metadata.mode = envModeMatch[2];
      continue;
    }

    // Parse [Isolation] Session: name
    const sessionMatch = line.match(/\[Isolation\] Session: (.+)/);
    if (sessionMatch) {
      metadata.session = sessionMatch[1];
      continue;
    }

    // Parse [Isolation] Image: name
    const imageMatch = line.match(/\[Isolation\] Image: (.+)/);
    if (imageMatch) {
      metadata.image = imageMatch[1];
      continue;
    }

    // Parse [Isolation] Endpoint: user@host
    const endpointMatch = line.match(/\[Isolation\] Endpoint: (.+)/);
    if (endpointMatch) {
      metadata.endpoint = endpointMatch[1];
      continue;
    }

    // Parse [Isolation] User: name (isolated)
    const userMatch = line.match(/\[Isolation\] User: (\w+)/);
    if (userMatch) {
      metadata.user = userMatch[1];
      continue;
    }
  }

  return metadata;
}

/**
 * Generate isolation metadata lines for timeline format
 * @param {object} metadata - Parsed isolation metadata
 * @param {string} [containerOrScreenName] - Container or screen session name
 * @returns {string[]} Array of timeline-formatted isolation lines
 */
function generateIsolationLines(metadata, containerOrScreenName = null) {
  const lines = [];

  if (metadata.isolation) {
    lines.push(createTimelineLine('isolation', metadata.isolation));
  }

  if (metadata.mode) {
    lines.push(createTimelineLine('mode', metadata.mode));
  }

  if (metadata.image) {
    lines.push(createTimelineLine('image', metadata.image));
  }

  // Use provided container/screen name or fall back to metadata.session
  if (metadata.isolation === 'docker') {
    const containerName = containerOrScreenName || metadata.session;
    if (containerName) {
      lines.push(createTimelineLine('container', containerName));
    }
  } else if (metadata.isolation === 'screen') {
    const screenName = containerOrScreenName || metadata.session;
    if (screenName) {
      lines.push(createTimelineLine('screen', screenName));
    }
  } else if (metadata.isolation === 'tmux') {
    const tmuxName = containerOrScreenName || metadata.session;
    if (tmuxName) {
      lines.push(createTimelineLine('tmux', tmuxName));
    }
  } else if (metadata.isolation === 'ssh') {
    if (metadata.endpoint) {
      lines.push(createTimelineLine('endpoint', metadata.endpoint));
    }
  }

  if (metadata.user) {
    lines.push(createTimelineLine('user', metadata.user));
  }

  return lines;
}

/**
 * Create a start block for command execution using timeline format
 * @param {object} options - Options for the block
 * @param {string} options.sessionId - Session UUID
 * @param {string} options.timestamp - Timestamp string
 * @param {string} options.command - Command being executed
 * @param {string[]} [options.extraLines] - Additional lines with isolation info
 * @param {boolean} [options.deferCommand] - If true, omit command line (for virtual command handling)
 * @param {string} [options.style] - Ignored (kept for backward compatibility)
 * @param {number} [options.width] - Ignored (kept for backward compatibility)
 * @returns {string} Formatted start block in timeline format
 */
function createStartBlock(options) {
  const {
    sessionId,
    timestamp,
    command,
    extraLines = [],
    deferCommand = false,
  } = options;

  const lines = [];

  // Header: session and start time
  lines.push(createTimelineLine('session', sessionId));
  lines.push(createTimelineLine('start', timestamp));

  // Parse and add isolation metadata if present
  const metadata = parseIsolationMetadata(extraLines);

  if (metadata.isolation) {
    lines.push(createEmptyTimelineLine());
    lines.push(...generateIsolationLines(metadata));
  }

  // Empty timeline line before command (always needed for separation)
  lines.push(createEmptyTimelineLine());

  // Command line (unless deferred for virtual command handling)
  if (!deferCommand) {
    lines.push(createCommandLine(command));
  }

  return lines.join('\n');
}

/**
 * Format duration in seconds with appropriate precision
 * @param {number} durationMs - Duration in milliseconds
 * @returns {string} Formatted duration string (e.g., "0.273s")
 */
function formatDuration(durationMs) {
  const seconds = durationMs / 1000;
  if (seconds < 0.001) {
    return '0.001s';
  } else if (seconds < 10) {
    // For durations under 10 seconds, show 3 decimal places
    return `${seconds.toFixed(3)}s`;
  } else if (seconds < 100) {
    return `${seconds.toFixed(2)}s`;
  } else {
    return `${seconds.toFixed(1)}s`;
  }
}

/**
 * Create a finish block for command execution using timeline format
 *
 * Bottom block ordering rules:
 * 1. Result marker (✓ or ✗)
 * 2. finish timestamp
 * 3. duration
 * 4. exit code
 * 5. (repeated isolation metadata, if any)
 * 6. empty timeline line
 * 7. log path (always second-to-last)
 * 8. session ID (always last)
 *
 * @param {object} options - Options for the block
 * @param {string} options.sessionId - Session UUID
 * @param {string} options.timestamp - Timestamp string
 * @param {number} options.exitCode - Exit code
 * @param {string} options.logPath - Path to log file
 * @param {number} [options.durationMs] - Duration in milliseconds
 * @param {string} [options.resultMessage] - Result message (ignored in new format)
 * @param {string[]} [options.extraLines] - Isolation info for repetition in footer
 * @param {string} [options.style] - Ignored (kept for backward compatibility)
 * @param {number} [options.width] - Ignored (kept for backward compatibility)
 * @returns {string} Formatted finish block in timeline format
 */
function createFinishBlock(options) {
  const {
    sessionId,
    timestamp,
    exitCode,
    logPath,
    durationMs,
    extraLines = [],
  } = options;

  const lines = [];

  // Result marker appears first in footer (after program output)
  lines.push(getResultMarker(exitCode));

  // Finish metadata
  lines.push(createTimelineLine('finish', timestamp));

  if (durationMs !== undefined && durationMs !== null) {
    lines.push(createTimelineLine('duration', formatDuration(durationMs)));
  }

  lines.push(createTimelineLine('exit', String(exitCode)));

  // Repeat isolation metadata if present
  const metadata = parseIsolationMetadata(extraLines);
  if (metadata.isolation) {
    lines.push(createEmptyTimelineLine());
    lines.push(...generateIsolationLines(metadata));
  }

  // Empty timeline line before final two entries
  lines.push(createEmptyTimelineLine());

  // Log and session are ALWAYS last (in that order)
  lines.push(createTimelineLine('log', logPath));
  lines.push(createTimelineLine('session', sessionId));

  return lines.join('\n');
}

/**
 * Escape a value for Links notation
 * Smart quoting: uses single or double quotes based on content
 * @param {string} str - String to escape
 * @returns {string} Escaped string
 */
function escapeForLinksNotation(str) {
  if (str === null || str === undefined) {
    return 'null';
  }

  const value = String(str);

  // Check for characters that need quoting
  const hasColon = value.includes(':');
  const hasDoubleQuotes = value.includes('"');
  const hasSingleQuotes = value.includes("'");
  const hasParens = value.includes('(') || value.includes(')');
  const hasNewline = value.includes('\n');
  const hasSpace = value.includes(' ');

  const needsQuoting =
    hasColon ||
    hasDoubleQuotes ||
    hasSingleQuotes ||
    hasParens ||
    hasNewline ||
    hasSpace;

  if (!needsQuoting) {
    return value;
  }

  if (hasDoubleQuotes && !hasSingleQuotes) {
    // Has " but not ' → use single quotes
    return `'${value}'`;
  } else if (hasSingleQuotes && !hasDoubleQuotes) {
    // Has ' but not " → use double quotes
    return `"${value}"`;
  } else if (hasDoubleQuotes && hasSingleQuotes) {
    // Has both " and ' → choose wrapper with fewer escapes
    const doubleQuoteCount = (value.match(/"/g) || []).length;
    const singleQuoteCount = (value.match(/'/g) || []).length;

    if (singleQuoteCount <= doubleQuoteCount) {
      // Escape single quotes by doubling them
      const escaped = value.replace(/'/g, "''");
      return `'${escaped}'`;
    } else {
      // Escape double quotes by doubling them
      const escaped = value.replace(/"/g, '""');
      return `"${escaped}"`;
    }
  } else {
    // Has colon, parentheses, newlines, or spaces but no quotes
    return `"${value}"`;
  }
}

/**
 * Format an object as nested Links notation
 * @param {object} obj - Object to format
 * @param {number} [indent=2] - Indentation level (spaces)
 * @param {number} [depth=0] - Current depth
 * @returns {string} Links notation formatted string
 */
function formatAsNestedLinksNotation(obj, indent = 2, depth = 0) {
  if (obj === null || obj === undefined) {
    return 'null';
  }

  if (typeof obj !== 'object') {
    return escapeForLinksNotation(obj);
  }

  if (Array.isArray(obj)) {
    // Format arrays
    if (obj.length === 0) {
      return '()';
    }
    const indentStr = ' '.repeat(indent * (depth + 1));
    const items = obj.map((item) => {
      const formatted = formatAsNestedLinksNotation(item, indent, depth + 1);
      return `${indentStr}${formatted}`;
    });
    return `(\n${items.join('\n')}\n${' '.repeat(indent * depth)})`;
  }

  // Format objects
  const entries = Object.entries(obj);
  if (entries.length === 0) {
    return '()';
  }

  const indentStr = ' '.repeat(indent * (depth + 1));
  const lines = entries
    .filter(([, value]) => value !== null && value !== undefined)
    .map(([key, value]) => {
      if (typeof value === 'object') {
        const nested = formatAsNestedLinksNotation(value, indent, depth + 1);
        return `${indentStr}${key}\n${nested}`;
      }
      const formattedValue = escapeForLinksNotation(value);
      return `${indentStr}${key} ${formattedValue}`;
    });

  return lines.join('\n');
}

module.exports = {
  // Timeline format API (formerly "status spine")
  TIMELINE_MARKER,
  SPINE, // deprecated, use TIMELINE_MARKER
  SUCCESS_MARKER,
  FAILURE_MARKER,
  createTimelineLine,
  createSpineLine, // deprecated, use createTimelineLine
  createEmptyTimelineLine,
  createEmptySpineLine, // deprecated, use createEmptyTimelineLine
  createCommandLine,
  getResultMarker,
  parseIsolationMetadata,
  generateIsolationLines,

  // Virtual command API (for multi-step execution)
  createVirtualCommandBlock,
  createVirtualCommandResult,
  createTimelineSeparator,

  // Main block creation functions
  createStartBlock,
  createFinishBlock,
  formatDuration,

  // Links notation utilities
  escapeForLinksNotation,
  formatAsNestedLinksNotation,
};

File: rust/src/lib/output_blocks.rs
//! Output formatting utilities for nicely rendered command blocks
//!
//! Provides "status spine" format: a width-independent, lossless output format
//! that works in TTY, tmux, SSH, CI, and logs.
//!
//! Core concepts:
//! - `│` prefix → tool metadata
//! - `$` → executed command
//! - No prefix → program output (stdout/stderr)
//! - Result marker (`✓` / `✗`) appears after output

use regex::Regex;

/// Metadata spine character
pub const SPINE: &str = "│";

/// Success result marker
pub const SUCCESS_MARKER: &str = "✓";

/// Failure result marker
pub const FAILURE_MARKER: &str = "✗";

/// Create a metadata line with spine prefix
pub fn create_spine_line(label: &str, value: &str) -> String {
    // Pad label to 10 characters for alignment
    format!("{} {:10}{}", SPINE, label, value)
}

/// Create an empty spine line (just the spine character)
pub fn create_empty_spine_line() -> String {
    SPINE.to_string()
}

/// Create a command line with $ prefix
pub fn create_command_line(command: &str) -> String {
    format!("$ {}", command)
}

/// Get the result marker based on exit code
pub fn get_result_marker(exit_code: i32) -> &'static str {
    if exit_code == 0 {
        SUCCESS_MARKER
    } else {
        FAILURE_MARKER
    }
}

/// Parsed isolation metadata
#[derive(Default)]
pub struct IsolationMetadata {
    pub isolation: Option<String>,
    pub mode: Option<String>,
    pub image: Option<String>,
    pub session: Option<String>,
    pub endpoint: Option<String>,
    pub user: Option<String>,
}

/// Parse isolation metadata from extra lines
pub fn parse_isolation_metadata(extra_lines: &[&str]) -> IsolationMetadata {
    let mut metadata = IsolationMetadata::default();

    let env_mode_re = Regex::new(r"\[Isolation\] Environment: (\w+), Mode: (\w+)").unwrap();
    let session_re = Regex::new(r"\[Isolation\] Session: (.+)").unwrap();
    let image_re = Regex::new(r"\[Isolation\] Image: (.+)").unwrap();
    let endpoint_re = Regex::new(r"\[Isolation\] Endpoint: (.+)").unwrap();
    let user_re = Regex::new(r"\[Isolation\] User: (\w+)").unwrap();

    for line in extra_lines {
        if let Some(caps) = env_mode_re.captures(line) {
            metadata.isolation = Some(caps[1].to_string());
            metadata.mode = Some(caps[2].to_string());
            continue;
        }

        if let Some(caps) = session_re.captures(line) {
            metadata.session = Some(caps[1].to_string());
            continue;
        }

        if let Some(caps) = image_re.captures(line) {
            metadata.image = Some(caps[1].to_string());
            continue;
        }

        if let Some(caps) = endpoint_re.captures(line) {
            metadata.endpoint = Some(caps[1].to_string());
            continue;
        }

        if let Some(caps) = user_re.captures(line) {
            metadata.user = Some(caps[1].to_string());
        }
    }

    metadata
}

/// Generate isolation metadata lines for spine format
pub fn generate_isolation_lines(
    metadata: &IsolationMetadata,
    container_or_screen_name: Option<&str>,
) -> Vec<String> {
    let mut lines = Vec::new();

    if let Some(ref isolation) = metadata.isolation {
        lines.push(create_spine_line("isolation", isolation));
    }

    if let Some(ref mode) = metadata.mode {
        lines.push(create_spine_line("mode", mode));
    }

    if let Some(ref image) = metadata.image {
        lines.push(create_spine_line("image", image));
    }

    // Use provided container/screen name or fall back to metadata.session
    if let Some(ref isolation) = metadata.isolation {
        let name = container_or_screen_name
            .map(String::from)
            .or_else(|| metadata.session.clone());

        if let Some(name) = name {
            match isolation.as_str() {
                "docker" => lines.push(create_spine_line("container", &name)),
                "screen" => lines.push(create_spine_line("screen", &name)),
                "tmux" => lines.push(create_spine_line("tmux", &name)),
                "ssh" => {
                    if let Some(ref endpoint) = metadata.endpoint {
                        lines.push(create_spine_line("endpoint", endpoint));
                    }
                }
                _ => {}
            }
        }
    }

    if let Some(ref user) = metadata.user {
        lines.push(create_spine_line("user", user));
    }

    lines
}

/// Options for creating a start block
pub struct StartBlockOptions<'a> {
    pub session_id: &'a str,
    pub timestamp: &'a str,
    pub command: &'a str,
    pub extra_lines: Option<Vec<&'a str>>,
    pub style: Option<&'a str>,
    pub width: Option<usize>,
}

/// Create a start block for command execution using status spine format
pub fn create_start_block(options: &StartBlockOptions) -> String {
    let mut lines = Vec::new();

    // Header: session and start time
    lines.push(create_spine_line("session", options.session_id));
    lines.push(create_spine_line("start", options.timestamp));

    // Parse and add isolation metadata if present
    if let Some(ref extra) = options.extra_lines {
        let metadata = parse_isolation_metadata(extra);

        if metadata.isolation.is_some() {
            lines.push(create_empty_spine_line());
            lines.extend(generate_isolation_lines(&metadata, None));
        }
    }

    // Empty spine line before command
    lines.push(create_empty_spine_line());

    // Command line
    lines.push(create_command_line(options.command));

    lines.join("\n")
}

/// Format duration in seconds with appropriate precision
pub fn format_duration(duration_ms: f64) -> String {
    let seconds = duration_ms / 1000.0;
    if seconds < 0.001 {
        "0.001s".to_string()
    } else if seconds < 10.0 {
        // For durations under 10 seconds, show 3 decimal places
        format!("{:.3}s", seconds)
    } else if seconds < 100.0 {
        format!("{:.2}s", seconds)
    } else {
        format!("{:.1}s", seconds)
    }
}

/// Options for creating a finish block
pub struct FinishBlockOptions<'a> {
    pub session_id: &'a str,
    pub timestamp: &'a str,
    pub exit_code: i32,
    pub log_path: &'a str,
    pub duration_ms: Option<f64>,
    pub result_message: Option<&'a str>,
    pub extra_lines: Option<Vec<&'a str>>,
    pub style: Option<&'a str>,
    pub width: Option<usize>,
}

/// Create a finish block for command execution using status spine format
///
/// Bottom block ordering rules:
/// 1. Result marker (✓ or ✗)
/// 2. finish timestamp
/// 3. duration
/// 4. exit code
/// 5. (repeated isolation metadata, if any)
/// 6. empty spine line
/// 7. log path (always second-to-last)
/// 8. session ID (always last)
pub fn create_finish_block(options: &FinishBlockOptions) -> String {
    let mut lines = Vec::new();

    // Result marker appears first in footer (after program output)
    lines.push(get_result_marker(options.exit_code).to_string());

    // Finish metadata
    lines.push(create_spine_line("finish", options.timestamp));

    if let Some(duration_ms) = options.duration_ms {
        lines.push(create_spine_line("duration", &format_duration(duration_ms)));
    }

    lines.push(create_spine_line("exit", &options.exit_code.to_string()));

    // Repeat isolation metadata if present
    if let Some(ref extra) = options.extra_lines {
        let metadata = parse_isolation_metadata(extra);
        if metadata.isolation.is_some() {
            lines.push(create_empty_spine_line());
            lines.extend(generate_isolation_lines(&metadata, None));
        }
    }

    // Empty spine line before final two entries
    lines.push(create_empty_spine_line());

    // Log and session are ALWAYS last (in that order)
    lines.push(create_spine_line("log", options.log_path));
    lines.push(create_spine_line("session", options.session_id));

    lines.join("\n")
}

/// Escape a value for Links notation
/// Smart quoting: uses single or double quotes based on content
pub fn escape_for_links_notation(value: &str) -> String {
    let has_colon = value.contains(':');
    let has_double_quotes = value.contains('"');
    let has_single_quotes = value.contains('\'');
    let has_parens = value.contains('(') || value.contains(')');
    let has_newline = value.contains('\n');
    let has_space = value.contains(' ');

    let needs_quoting = has_colon
        || has_double_quotes
        || has_single_quotes
        || has_parens
        || has_newline
        || has_space;

    if !needs_quoting {
        return value.to_string();
    }

    if has_double_quotes && !has_single_quotes {
        // Has " but not ' → use single quotes
        format!("'{}'", value)
    } else if has_single_quotes && !has_double_quotes {
        // Has ' but not " → use double quotes
        format!("\"{}\"", value)
    } else if has_double_quotes && has_single_quotes {
        // Has both " and ' → choose wrapper with fewer escapes
        let double_quote_count = value.matches('"').count();
        let single_quote_count = value.matches('\'').count();

        if single_quote_count <= double_quote_count {
            // Escape single quotes by doubling them
            let escaped = value.replace('\'', "''");
            format!("'{}'", escaped)
        } else {
            // Escape double quotes by doubling them
            let escaped = value.replace('"', "\"\"");
            format!("\"{}\"", escaped)
        }
    } else {
        // Has colon, parentheses, newlines, or spaces but no quotes
        format!("\"{}\"", value)
    }
}

/// Format a serde_json::Value for Links notation
pub fn format_value_for_links_notation(value: &serde_json::Value) -> String {
    match value {
        serde_json::Value::Null => "null".to_string(),
        serde_json::Value::Bool(b) => b.to_string(),
        serde_json::Value::Number(n) => n.to_string(),
        serde_json::Value::String(s) => escape_for_links_notation(s),
        serde_json::Value::Array(_) | serde_json::Value::Object(_) => {
            // For complex types, serialize and quote
            let s = serde_json::to_string(value).unwrap_or_default();
            escape_for_links_notation(&s)
        }
    }
}
